src/js/helpers/timeline/declarationHelper.js

```diff
@@ -0,0 +1,119 @@
+
+const getNextNodes = (node) => node.on[node.response.transition];
+
+const hasProcessedTransition = (node) => !!node.on && !!node.response && !!node.response.transition
+  && !!getNextNodes(node);
+
+export const generateUniqueLabel = (label, takenLabels) => {
+  const usedLabels = takenLabels;
+
+  if (usedLabels[label]) {
+    usedLabels[label] += 1;
+    return `${label} (#${usedLabels[label]})`;
+  }
+
+  usedLabels[label] = 1;
+  return label;
+};
+
+export const getGraphWithUniqueLabels = (root, takenLabels = {}) => ({
+  ...root,
+  uniqueLabel: generateUniqueLabel(root.label, takenLabels),
+  ...(root.on
+    ? {
+      on: Object.entries(root.on)
+        .map(([transition, node]) => [transition, getGraphWithUniqueLabels(node, takenLabels)])
+        .reduce((total, [transition, node]) => ({ ...total, [transition]: node }), {}),
+    }
+    : {}),
+  ...(root.subtasks
+    ? { subtasks: (root.subtasks).map((node) => getGraphWithUniqueLabels(node, takenLabels)) }
+    : {}),
+});
+
+export const getProcessedNodes = (root, generateUniqueLabels = true) => {
+  const rootNode = generateUniqueLabels ? getGraphWithUniqueLabels(root) : root;
+
+  return (hasProcessedTransition(rootNode)
+    ? [rootNode, ...getProcessedNodes(getNextNodes(rootNode), false)]
+    : [rootNode]);
+};
+
+const generateGroupName = (node) => node.uniqueLabel;
+
+const generateSubgroupData = (parent) => (parent ? { subgroupOf: generateGroupName(parent) } : {});
+
+const createTimelineItem = (node, parent) => ({
+  id: node.id,
+  start: parseInt(node.started, RADIX),
+  end: parseInt(node.finished, RADIX),
+  content: EMPTY_LABEL,
+  group: generateGroupName(node),
+  ...generateSubgroupData(parent),
+});
+
+export const createTimelineItems = (node, parent = null) => (hasSubtasks(node)
+  ? [
+    createTimelineItem(node, parent),
+    ...node.subtasks
+      .flatMap((subtask) => getProcessedNodes(subtask))
+      .flatMap((subtask) => createTimelineItems(subtask, node))]
+  : [
+    createTimelineItem(node, parent),
+  ]);
+
+export const reduceSubgroups = (prev, { group, subgroupOf }) => {
+  const total = { ...prev };
+
+  if (subgroupOf) {
+    total[subgroupOf] = [...total[subgroupOf], group];
+  }
+
+  total[group] = total[group] || [];
+
+  return total;
+};
+
+export const extractGroupData = (items) => Object.entries(items
+  .reduce(reduceSubgroups, {}))
+  .map(([key, value]) => ({ name: key, subgroups: value }));
+
+const createTimelineGroup = (name, index, subgroups) => ({
+  id: name,
+  order: index,
+  content: name,
+  nestedGroups: subgroups,
+});
+
+export const constructTimeline = (json) => {
+  const items = getProcessedNodes(json)
+    .flatMap((node) => createTimelineItems(node));
+
+  const groups = extractGroupData(items)
+    .map((group) => (group.subgroups.length > 0 ? group : { ...group, subgroups: null }))
+    .map((group, index) => createTimelineGroup(group.name, index, group.subgroups));
```

Comment:
I believe it could be simplified to:
```
.map(({name, subgroups}, index) => createTimelineGroup(name, index, subgroups. length ? subgroups : null));
```

Reply:
Nice!